cipher: add encrypt/decrypt-only block cipher traits

tarcieri opened the pull request:
Closes #349

Adds `BlockEncrypt` and `BlockDecrypt` traits which can be used in cases where e.g. only the encryption portion of a block cipher is used, as in CTR mode.

This PR does not otherwise attempt to add things like a blanket impl.

tarcieri commented:
Note: I used the new suggested method names here: `*_block`, `*_par_blocks`, `*_blocks`.

Perhaps it would make more sense to leave them the same for consistency.

tarcieri commented on cipher/src/block.rs:
```diff
@@ -185,3 +185,97 @@ impl<Alg: BlockCipher> BlockCipher for &Alg {
+    /// Encrypt a slice of blocks, leveraging parallelism when available.
+    #[inline]
+    fn encrypt_blocks(&self, mut blocks: &mut [GenericArray<u8, Self::BlockSize>]) {
+        let pb = Self::ParBlocks::to_usize();
+
+        if pb > 1 {
+            let mut iter = blocks.chunks_exact_mut(pb);
+
+            for chunk in &mut iter {
+                self.encrypt_par_blocks(chunk.try_into().unwrap())
+            }
+
+            blocks = iter.into_remainder();
+        }
+
+        for block in blocks {
+            self.encrypt_block(block);
+        }
```
This is copypasta from the `BlockCipher` trait.

It may make more sense to extract some sort of common method which encapsulates this behavior.

tarcieri commented on cipher/src/block.rs:
```diff
@@ -185,3 +185,97 @@ impl<Alg: BlockCipher> BlockCipher for &Alg {
+    /// Decrypt a slice of blocks, leveraging parallelism when available.
+    #[inline]
+    fn decrypt_blocks(&self, mut blocks: &mut [GenericArray<u8, Self::BlockSize>]) {
+        let pb = Self::ParBlocks::to_usize();
+
+        if pb > 1 {
+            let mut iter = blocks.chunks_exact_mut(pb);
+
+            for chunk in &mut iter {
+                self.decrypt_blocks(chunk.try_into().unwrap())
+            }
+
+            blocks = iter.into_remainder();
+        }
+
+        for block in blocks {
+            self.decrypt_block(block);
+        }
```
More copypasta here.

tarcieri commented:
After a few attempts at eliminating the redundancy, I think it might make more sense to transform the `BlockCipher` trait (via a breaking change) into a marker trait for the case where a type impls both `Encrypt` and `Decrypt` with compatible `BlockSize`/`ParBlocks`.

newpavlov reviewed (commented):
Aren't `Encrypt`/`Decrypt` too generic? It could be worth to use something like `BlockEncrypt` instead of `block::Encrypt`. Such naming would also allow us to simplify the crate structure by removing `stream` and `block` modules.

tarcieri commented on cipher/src/block.rs:
```diff
@@ -185,3 +185,97 @@ impl<Alg: BlockCipher> BlockCipher for &Alg {
+/// Encrypt-only functionality for block ciphers
+pub trait Encrypt {
+    /// Size of the block in bytes
+    type BlockSize: ArrayLength<u8>;
```
Yeah, that sounds good

tarcieri commented:
>  It could be worth to use something like BlockEncrypt instead of block::Encrypt. Such naming would also allow us to simplify the crate structure by removing stream and block modules.

That's a potential alternative, yes. I think it makes sense for the code to still be factored into two modules, but we could re-export everything at the toplevel rather than using modules.

tarcieri commented:
In 5096a41 I renamed the traits to `BlockEncrypt` and `BlockDecrypt`, and added a `BlockSizeMarker` trait for carrying the `BlockSize` associated type/constant.

newpavlov commented:
>I think it makes sense for the code to still be factored into two modules

I agree, I meant how API will look on outside.

tarcieri commented:
@newpavlov does this look ok now, or would you rather hold off until `cipher` v0.3?

newpavlov commented:
I think it will be better to release these changes in v0.3.

tarcieri commented:
@newpavlov I've gone ahead and made the discussed breaking changes in 2c61ca2, including splitting `BlockCipherMut` into `BlockEncryptMut` and `BlockDecryptMut` which require the given type impls `BlockCipher`.

This had the added benefit of eliminating the need for a `stream::FromBlockCipherMut` trait.
